python_ta/debug/snapshot_tracer.py

```diff
@@ -39,37 +53,119 @@ def __init__(self, output_directory: Optional[str] = None, **kwargs) -> None:
     def __exit__(self, exc_type, exc_val, exc_tb) -> None:
-        """Remove the trace function."""
+        """Remove the trace function and run any post-tracing functions."""
         sys.settrace(None)
-        inspect.getouterframes(inspect.currentframe())[1].frame.f_trace = None
+        self._func_frame.f_trace = None
+        if self.open_webstepper:
+            self._build_result_html()
+            self._open_html()
+
+    def _build_result_html(self) -> None:
+        """Build and write the Webstepper html to the output directory"""
+        current_dir = os.path.dirname(os.path.abspath(__file__))
+
+        html_content = self._read_original_html(current_dir)
+        soup = BeautifulSoup(html_content, "html.parser")
+
+        self._modify_bundle_import_path(current_dir, soup)
+        self._insert_data(soup)
+        self._write_generated_html(soup)
+
+    def _open_html(self) -> None:
+        """Open the generated HTML file in a web browser."""
+        index_html = f"file://{os.path.join(self.output_directory, 'index.html')}"
+        webbrowser.open(index_html, new=2)
+
+    def _read_original_html(self, current_dir: str) -> None:
+        """Read the original Webstepper html"""
+        original_index_html = os.path.join(current_dir, "webstepper", "index.html")
+        with open(original_index_html, "r") as file:
+            html_content = file.read()
+        return html_content
+
+    def _modify_bundle_import_path(self, current_dir: str, soup: BeautifulSoup) -> None:
+        """Modify the bundle path to the absolute path to the bundle"""
+        original_js_bundle = os.path.join(current_dir, "webstepper", "index.bundle.js")
+        soup.select("script")[0]["src"] = f"file://{original_js_bundle}"
+
+    def _insert_data(self, soup: BeautifulSoup) -> None:
+        """Insert the SVG array and code string into the Webstepper index HTML."""
+        code_script = f"<script>window.codeText=`{self._get_code()}` </script>\n"
+        svg_script = f"<script>window.svgArray={json.dumps(self._snapshots)}</script>\n"
+        soup.select("script")[0].insert_before(BeautifulSoup(code_script, "html.parser"))
+        soup.select("script")[0].insert_before(BeautifulSoup(svg_script, "html.parser"))
+
+    def _write_generated_html(self, soup: BeautifulSoup) -> None:
+        """Write the generated Webstepper html to the output directory"""
+        modified_index_html = os.path.join(self.output_directory, "index.html")
+        with open(modified_index_html, "w") as file:
+            file.write(str(soup))
+
+    def _get_code(self) -> str:
+        """Retrieve and save the code string to be displayed in Webstepper."""
+        code_string = inspect.cleandoc(inspect.getsource(self._func_frame))
+        i = self._first_line - self._func_frame.f_code.co_firstlineno
+        lst_str_lines = code_string.splitlines()
+        lst_from_with_stmt = lst_str_lines[i:]
+
+        num_whitespace = 0
+        for char in lst_from_with_stmt[0]:
```

Comment:
don't use a for loop; instead compare the length against the string with `lstrip` called